packages/flutter_tools/lib/src/runner/local_engine.dart

```diff
@@ -113,12 +76,79 @@ class LocalEngineLocator {
     if (engineSourcePath != null) {
       _logger.printTrace('Local engine source at $engineSourcePath');
       return _findEngineBuildPath(localEngine, engineSourcePath);
-    } else if (localEngine != null) {
-      _logger.printTrace('Could not find engine source for $localEngine, skipping');
+    }
+    if (localEngine != null) {
+      throwToolExit(
+        _userMessages.runnerNoEngineSrcDir(
+          kFlutterEnginePackageName,
+          kFlutterEngineEnvironmentVariableName,
+        ),
+        exitCode: 2,
+      );
+    }
+    return null;
+  }
+
+  String _findEngineSourceByLocalEngine(String localEngine) {
+    // When the local engine is an absolute path to a variant inside the
+    // out directory, parse the engine source.
+    // --local-engine /path/to/cache/builder/src/out/host_debug_unopt
+    if (_fileSystem.path.isAbsolute(localEngine)) {
+      final Directory localEngineDirectory = _fileSystem.directory(localEngine);
+      final Directory outDirectory = localEngineDirectory?.parent;
+      final Directory srcDirectory = outDirectory?.parent;
+      if (localEngineDirectory.existsSync() && outDirectory.basename == 'out' && srcDirectory.basename == 'src') {
+        _logger.printTrace('Parsed engine source from local engine as ${srcDirectory.path}.');
+        return srcDirectory.path;
+      }
     }
     return null;
   }
 
+  Future<String> _findEngineSourceBySkyEngine(String packagePath) async {
```

Comment:
This method was extracted from `findEnginePath` to make that method more readable, but has not been changed.